### PR TITLE
Add support for reloading static files

### PR DESCRIPTION
**Pull request by pranavkm**
* Update vanilla css files without a browser reload
* Refresh the browser when changes to other static content is detected.

### CODE REVIEW

File: src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs [hunk 1]
```diff
@@ -86,23 +91,43 @@ public async Task<IFileSet> CreateAsync(CancellationToken cancellationToken)
...
+
+                        commandLine.Execute(lines);
+                        var isNetCoreApp31OrNewer = isNetCoreApp31.Value();
+                        var fileItems = new List<FileItem>();
+                        foreach (var file in files.Values)
+                        {
+                            fileItems.Add(new FileItem(file));
+                        }
+
+                        // Ignore static files if we do not support it.
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs:111]:
```suggestion
```

File: src/Tools/dotnet-watch/src/Program.cs [hunk 2]
```diff
@@ -202,9 +207,24 @@ private void OnCancelKeyPress(object sender, ConsoleCancelEventArgs args)
                 return 1;
             }
 
-            foreach (var file in files)
+            foreach (var group in files.GroupBy(g => g.FileKind).OrderBy(g => g.Key))
             {
-                _console.Out.WriteLine(file);
+                if (group.Key == FileKind.StaticFile)
+                {
+                    _console.Out.WriteLine($"::: Watch Action: Refresh browser :::");
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/src/Program.cs:214]:
```suggestion
                    _console.Out.WriteLine("::: Watch Action: Refresh browser :::");
```

File: src/Tools/dotnet-watch/BrowserRefresh/src/WebSocketScriptInjection.js [hunk 3]
```diff
@@ -8,15 +8,55 @@ setTimeout(function () {
...
+        location.reload();
+        return;
+      }
+
+      const styleElement = document.querySelector(`link[href^="${fileName}"]`) ||
+        document.querySelector(`link[href^="${document.baseURI}${fileName}"]`);
+      if (styleElement && styleElement.parentNode) {
+        if (styleElement.loading) {
+          // A file change notification may be triggered for the same file before the browser
+          // finishes processing a previous update. In this case, it's easiest to igore later updates
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/BrowserRefresh/src/WebSocketScriptInjection.js:35]:
```suggestion
          // finishes processing a previous update. In this case, it's easiest to ignore later updates
```

File: src/Tools/dotnet-watch/src/DotNetWatcher.cs [hunk 4]
```diff
@@ -93,15 +94,38 @@ public async Task WatchAsync(ProcessSpec processSpec, CancellationToken cancella
...
                     var args = ArgumentEscaper.EscapeAndConcatenate(processSpec.Arguments);
                     _reporter.Verbose($"Running {processSpec.ShortDisplayName()} with the following arguments: {args}");
 
                     _reporter.Output("Started");
 
-                    var finishedTask = await Task.WhenAny(processTask, fileSetTask, cancelledTaskSource.Task);
+                    bool waitForFileChanges;
+                    Task<FileItem?> fileSetTask;
+                    Task finishedTask;
+                    do
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/DotNetWatcher.cs:106]: Can't this just be:
```c#

                 while (true) {
                        fileSetTask = fileSetWatcher.GetChangedFileAsync(combinedCancellationSource.Token);
                        finishedTask = await Task.WhenAny(processTask, fileSetTask, cancelledTaskSource.Task);

                        if (context.BrowserRefreshServer is not null &&
                            finishedTask == fileSetTask &&
                            fileSetTask.Result is FileItem { FileKind: FileKind.StaticFile } file)
                        {
                            _reporter.Verbose($"Handling file change event for static content {file.FilePath}.");

                            // If we can handle the file change without a browser refresh, do it.
                            await StaticContentHandler.TryHandleFileAction(context.BrowserRefreshServer, file, combinedCancellationSource.Token);
                        }
                        else
                        {
                            break;
                        }
                    }
```

File: src/Tools/dotnet-watch/src/DotNetWatcher.cs [hunk 5]
```diff
@@ -125,26 +149,27 @@ public async Task WatchAsync(ProcessSpec processSpec, CancellationToken cancella
...
+                        _reporter.Output($"File changed: {changedFile.Value.FilePath}");
                     }
                 }
             }
         }
 
         public async ValueTask DisposeAsync()
         {
-            foreach (var filter in _filters)
+           foreach (var filter in _filters)
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/DotNetWatcher.cs:172]: nit undo

File: src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs [hunk 6]
```diff
@@ -74,7 +78,8 @@ public async Task<IFileSet> CreateAsync(CancellationToken cancellationToken)
                             "msbuild",
                             "/nologo",
                             _projectFile,
-                            $"/p:_DotNetWatchListFile={watchList}"
+                            $"/p:_DotNetWatchListFile={watchList}",
+                            _dotNetWatchOptions.SuppressHandlingStaticContentFiles ? "/p:DotNetWatchContentFiles=false" : "",
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs:82]: Seem odd to have the env var (SUPPRESS) default to false while the property passed in default to true.

File: src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs [hunk 7]
```diff
@@ -86,23 +91,42 @@ public async Task<IFileSet> CreateAsync(CancellationToken cancellationToken)
...
                         var lines = File.ReadAllLines(watchList);
-                        var isNetCoreApp31OrNewer = lines.FirstOrDefault() == "true";
 
-                        var fileset = new FileSet(
-                            isNetCoreApp31OrNewer,
-                            lines.Skip(1)
-                                .Select(l => l?.Trim())
-                                .Where(l => !string.IsNullOrEmpty(l)));
+                        var staticFiles = new List<string>();
+                        var commandLine = new CommandLineApplication();
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs:96]: Context for this change? Confused why you are creating a CommandLineApplication here.

File: src/Tools/dotnet-watch/src/StaticContentHandler.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+namespace Microsoft.DotNet.Watcher.Tools
+{
+    public class StaticContentHandler
+    {
+        private static readonly byte[] UpdateCssMessage = Encoding.UTF8.GetBytes("UpdateStaticFile||");
+
+        internal static async ValueTask TryHandleFileAction(BrowserRefreshServer browserRefreshServer, FileItem fileItem, CancellationToken cancellationToken)
+        {
+            var filePath = fileItem.StaticWebAssetPath;
+            var bytesToRent = UpdateCssMessage.Length + Encoding.UTF8.GetMaxByteCount(filePath.Length + 2);
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/StaticContentHandler.cs:19]: nit: Why + 2?

File: src/Tools/dotnet-watch/src/Program.cs [hunk 9]
```diff
@@ -202,9 +207,24 @@ private void OnCancelKeyPress(object sender, ConsoleCancelEventArgs args)
                 return 1;
             }
 
-            foreach (var file in files)
+            foreach (var group in files.GroupBy(g => g.FileKind).OrderBy(g => g.Key))
             {
-                _console.Out.WriteLine(file);
+                if (group.Key == FileKind.StaticFile)
+                {
+                    _console.Out.WriteLine("::: Watch Action: Refresh browser :::");
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/Program.cs:214]: What is this for?

File: src/Tools/dotnet-watch/src/StaticContentHandler.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+            var bytes = ArrayPool<byte>.Shared.Rent(bytesToRent);
+
+            try
+            {
+                UpdateCssMessage.CopyTo(bytes.AsSpan());
+                var length = UpdateCssMessage.Length;
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/src/StaticContentHandler.cs:25]: nit: could you make the precalculated length a const or at least make it so you aren't accessing the length property twice?

File: src/Tools/dotnet-watch/test/FileWatcherTests.cs [hunk 11]
```diff
@@ -68,7 +68,7 @@ public async Task ChangeFile(bool usePolling)
                     var changedEv = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
                     var filesChanged = new HashSet<string>();
 
-                    EventHandler<string> handler = null;
```
Comment (jkotalik) [inline src/Tools/dotnet-watch/test/FileWatcherTests.cs:71]: What's your opinion on nullable for tests?

**Review by jkotalik — APPROVED** [review]
Little wary of the js code to send file change notifications for css, but seems good besides that.

File: src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs (hunk 6 above)
Comment (pranavkm) [reply src/Tools/dotnet-watch/src/Internal/MsBuildFileSetFactory.cs:82]: DotNetWatch has precedence for using BLAH_SUPPRESS_FEATURE pattern to turn off features: https://docs.microsoft.com/en-us/aspnet/core/tutorials/dotnet-watch?view=aspnetcore-5.0#dotnet-watch-configuration. MSBuild does not typically do this. You usually set a property to false to turn it off. We want both ways to be work (turn off file watching for content files on a per project basis and turn it off as a watch feature).

I'm open to suggestions though.

**Pull request merged**